File: sol/solution/0001-0100/0032. Longest Valid Parentheses/Solution.ex
defmodule Solution do
  @spec longest_valid_parentheses(String.t()) :: integer
  def longest_valid_parentheses(s) do
    lr = pass_lr(s)
    rl = pass_rl(s)
    max(lr, rl)
  end
  defp pass_lr(s) do
    s
    |> String.graphemes()
    |> Enum.reduce({0, 0, 0}, fn ch, {left, right, best} ->
      {l, r} = if ch == "(", do: {left + 1, right}, else: {left, right + 1}
      cond do
        l == r -> {l, r, max(best, 2 * r)}
        r > l  -> {0, 0, best}
        true   -> {l, r, best}
      end
    end)
    |> elem(2)
  end
  defp pass_rl(s) do
    s
    |> String.graphemes()
    |> Enum.reverse()
    |> Enum.reduce({0, 0, 0}, fn ch, {left, right, best} ->
      {l, r} = if ch == ")", do: {left, right + 1}, else: {left + 1, right}
      cond do
        l == r -> {l, r, max(best, 2 * l)}
        l > r  -> {0, 0, best}
        true   -> {l, r, best}
      end
    end)
    |> elem(2)
  end
end

File: sol/solution/0001-0100/0032. Longest Valid Parentheses/Solution.rb
# @param {String} s
# @return {Integer}
def longest_valid_parentheses(s)
  ans = left = right = 0
  s.each_char do |ch|
    if ch == '(' then left += 1 else right += 1 end
    if left == right
      ans = [ans, 2 * right].max
    elsif right > left
      left = right = 0
    end
  end
  left = right = 0
  s.reverse.each_char do |ch|
    if ch == ')' then right += 1 else left += 1 end
    if left == right
      ans = [ans, 2 * left].max
    elsif left > right
      left = right = 0
    end
  end
  ans
end